File: RB101_Programming_Foundations/Small Probems/Easy8/Python/palindromic_substrings.py



'''Write a method that returns a list of all substrings of a string that are palindromic. That is, each substring must consist of the same sequence of characters forwards as it does backwards. The return value should be arranged in the same sequence as the substrings appear in the string. Duplicate palindromes should be included multiple times.

You may (and should) use the substrings method you wrote in the previous exercise.

For the purposes of this exercise, you should consider all characters and pay attention to case; that is, "AbcbA" is a palindrome, but neither "Abcba" nor "Abc-bA" are. In addition, assume that single characters are not palindromes.
'''

'''
Input: string

Output: array of palindromic substrings

Rules


Data structure / Algorithms
    I can call the substrings function to get all possible substrings from the input string
    create an array for storing palindromic substrings
    iterate through the substrings array
        check if current substring is palindromic
        if it is, add it to the array
    return the palindromic array

'''

def palindromes(string):
    substring_array = substrings(string)

    palindromes = []

    for substring in substring_array:
        if substring[::-1] == substring and len(substring) > 1:
            palindromes.append(substring)
    return palindromes








def substrings(string):
    result_array = []

    for idx in range(0, len(string)):
        result_array += substrings_at_start(string[idx:])
    return result_array




def substrings_at_start(string):
    result_array = [""] * len(string)

    for idx in range(0, len(string)):
        result_array[idx] = string[:idx + 1]
    return result_array


print(palindromes('abcd') == [])
print(palindromes('madam') == ['madam', 'ada'])
print(palindromes('hello-madam-did-madam-goodbye') == [
  'll', '-madam-', '-madam-did-madam-', 'madam', 'madam-did-madam', 'ada',
  'adam-did-mada', 'dam-did-mad', 'am-did-ma', 'm-did-m', '-did-', 'did',
  '-madam-', 'madam', 'ada', 'oo'
])
print(palindromes('knitting cassettes') == [
  'nittin', 'itti', 'tt', 'ss', 'settes', 'ette', 'tt'
])

File: RB101_Programming_Foundations/small_problems/Easy8/palindromic_substrings.rb
# input : string
# output : an array of strings that are all possible substrings to the original that are palindromic
# palindrome is a word that is the same read forward and reverse



def palindromes(string)
  all_substrings = substrings(string)
  all_substrings.select { |word| palindrome?(word) }
end

def palindrome?(string)
  #ignores non-alphanumeric and is case insensitive
  string = string.delete('^A-Za-z0-9').downcase
  string == string.reverse && string.length > 1

end


def substrings(string)
  array = []
  string_length = string.size

  string_length.times do |idx|
    (string_length - idx).times do |last_index|
      array << string[idx..(idx + last_index)]
    end
  end

  array
end


p palindromes('abcd') == []
p palindromes('madam') == ['madam', 'ada']
p palindromes('hello-madam-did-madam-goodbye') == [
  'll', '-madam-', '-madam-did-madam-', 'madam', 'madam-did-madam', 'ada',
  'adam-did-mada', 'dam-did-mad', 'am-did-ma', 'm-did-m', '-did-', 'did',
  '-madam-', 'madam', 'ada', 'oo'
]
p palindromes('knitting cassettes') == [
  'nittin', 'itti', 'tt', 'ss', 'settes', 'ette', 'tt'
]